### ISSUE TITLE
Map Hanging With Generator

### ISSUE THREAD

**Issue by tkram01** [PrefectHQ/prefect#7299]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

When running the below code Prefect hangs and never completes. There is no error thrown. The code works fine with an n of 5 or 100.

### Reproduction

```python3
from prefect import flow, task


@task
def task1(x: int) -> int:
    return x + 10


@task
def task2(x: int) -> int:
    return -x


@flow()
def run_my_flow(n: int):
    task2.map(task1.map(range(n)))


if __name__ == "__main__":
    n = 500
    print(run_my_flow(n))
```

### Versions

```Text
Version:             2.6.4
API version:         0.8.2
Python version:      3.10.6
Git commit:          51e92dda
Built:               Thu, Oct 20, 2022 3:11 PM
OS/Arch:             darwin/x86_64
Profile:             workflow2
Server type:         cloud
```

**Comment by jacobdanovitch** [comment]
This may not be specific to generators, I can reproduce this with `list(range(n))` as well. The flow runs for a while, submitting and completing some number of tasks, and then hangs at:

```
18:28:41.699 | DEBUG   | prefect.client - Connecting to API at http://orion:4200/api/
```

Results for various task runners:

* `RayTaskRunner`: **crashes** with exception `BrokenPipeError: [Errno 32] Broken pipe`, flow run uses 40-80% CPU and 7% memory, orion 40-70% CPU, postgres 60-100% CPU
* `DaskTaskRunner`: **hangs**, flow run uses 40-80% CPU and 6% memory, orion 60-110%, postgres 60-100%
* `ConcurrentTaskRunner`: **hangs**, flow run 40% and 1% memory, postgres 40-100%, orion 40-80%
* `SequentialTaskRunner`: **succeeds**, flow run 40-70%, oroion 40-70%, postgres 50-100%

In all three asynchronous task runners, the memory and CPU usage of the flow run eventually go to 0. In all cases, the Orion server and Postgres use around 1% memory. Setting `PREFECT_API_REQUEST_TIMEOUT='60.0'` in the flow run and `PREFECT_ORION_DATABASE_CONNECTION_TIMEOUT='10'` (and `'45'`) in Orion had no effect.

This seems to be the result whenever I run a large-ish number of tasks (>=~200).  I'm in a docker setup using Postgres but I see this kind of behavior locally and on Kubernetes.

**Comment by madkinsz** [comment]
@jacobdanovitch Thanks for the detailed notes! We believe this is caused by the default AnyIO thread capacity limit of 40 which gets consumed by tasks waiting for upstream results. We are exploring a design to improve this

**Comment by jacobdanovitch** [comment]
@madkinsz Gotcha, makes sense. Is there an ETA on that and/or a workaround in the mean time?

**Comment by madkinsz** [comment]
It's my highest priority, can't promise a timeline though as we're not certain of the root cause. You could bump the count: https://github.com/agronholm/anyio/blob/master/docs/threads.rst#adjusting-the-default-maximum-worker-thread-count it'll degrade performance at some point so we need to find a way to have waiting tasks release threads.

**Comment by multimeric** [comment]
Interestingly I get it hanging even using the `SequentialTaskRunner`. In terms of priority, this seems like a big issue to me, because it means that workflows with > 1000 branches are completely broken on Prefect. Am I wrong?

**Comment by jacobdanovitch** [comment]
> Interestingly I get it hanging even using the `SequentialTaskRunner`. In terms of priority, this seems like a big issue to me, because it means that workflows with > 1000 branches are completely broken on Prefect. Am I wrong?

What kind of machine are you using? I'm able to run upwards of 4k tasks with the Sequential runner just fine on 4 CPUs & 16GB RAM.

**Comment by multimeric** [comment]
Admittedly on a tiny machine, but the tasks were just print statements so I thought it might work.

**Comment by madkinsz** [comment]
Resolved by #8887
